subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java

```diff
@@ -56,31 +56,51 @@ public DefaultDeleter(LongSupplier timeProvider, Predicate<? super File> isSymli
     @Override
     public boolean deleteRecursively(File root, boolean followSymlinks) throws IOException {
         if (root.exists()) {
-            LOGGER.debug("Deleting {}", root);
-            long startTime = timeProvider.getAsLong();
-            Collection<String> failedPaths = new ArrayList<String>();
-            deleteRecursively(startTime, root, root, followSymlinks, failedPaths);
-            if (!failedPaths.isEmpty()) {
-                throwWithHelpMessage(startTime, root, followSymlinks, failedPaths, false);
-            }
-            return true;
+            return deleteRecursively(root, followSymlinks, false);
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean cleanRecursively(File root, boolean followSymlinks) throws IOException {
+        if (root.exists() && shouldRemoveContentsOf(root, followSymlinks)) {
```

Comment:
I think for `cleanRecursively`, we should always delete the contents of the root, even if it is a symlink. If we don't, then the directory won't be empty after this call. We can leave this until we make the general change whether to follow symlinks for deletion or not.

Reply:
I've introduced `ensureEmptyDirectory()`, PTAL.